// e3db-client-interface/dist/types/clientDetails.js
/*!
 * Tozny e3db
 *
 * LICENSE
 *
 * Tozny dual licenses this product. For commercial use, please contact
 * info@tozny.com. For non-commercial use, the contents of this file are
 * subject to the TOZNY NON-COMMERCIAL LICENSE (the "License") which
 * permits use of the software only by government agencies, schools,
 * universities, non-profit organizations or individuals on projects that
 * do not receive external funding other than government research grants
 * and contracts.  Any other use requires a commercial license. You may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at https://tozny.com/legal/non-commercial-license.
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations under
 * the License. Portions of the software are Copyright (c) TOZNY LLC, 2018-19.
 * All rights reserved.
 *
 * @copyright Copyright (c) 2018-19 Tozny, LLC (https://tozny.com)
 */
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _publicKey = _interopRequireDefault(require("./publicKey"));

var _signingKey = _interopRequireDefault(require("./signingKey"));

/**
 * Full information about a specific E3DB client, including the client's
 * public/private keys for cryptographic operations and API credentials.
 *
 * @property {string}     clientId   UUID representing the client.
 * @property {string}     apiKeyId   API key to be used when authenticating with e3db
 * @property {string}     apiSecret  API password to be used when authenticating with e3db
 * @property {PublicKey}  publicKey  Curve25519 public key for the client.
 * @property {string}     name       Description of the client
 * @property {SigningKey} signingKey Ed25519 public key for the client.
 */
var ClientDetails =
/*#__PURE__*/
function () {
  function ClientDetails(clientId, apiKeyId, apiSecret, publicKey, name) {
    var signingKey = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : null;
    (0, _classCallCheck2["default"])(this, ClientDetails);
    this.clientId = clientId;
    this.apiKeyId = apiKeyId;
    this.apiSecret = apiSecret;
    this.publicKey = publicKey;
    this.name = name;

    if (signingKey === null) {
      signingKey = new _signingKey["default"](null);
    }

    this.signingKey = signingKey;
  }
  /**
   * Specify how an already unserialized JSON array should be marshaled into
   * an object representation.
   *
   * Client information contains the ID of the client, API credentials for interacting
   * with the e3db server, a Curve25519 public key component, and a description of the
   * client as specified during creation.
   *
   * <code>
   * info = ClientDetails::decode({
   *   client_id: '',
   *   api_key_id: '',
   *   api_secret: '',
   *   public_key: {
   *     curve25519: ''
   *   },
   *   signing_key: {
   *     ed25519: ''
   *   },
   *   name: ''
   * })
   * <code>
   *
   * @param {object} json
   *
   * @return {Promise<ClientDetails>}
   */


  (0, _createClass2["default"])(ClientDetails, null, [{
    key: "decode",
    value: function () {
      var _decode = (0, _asyncToGenerator2["default"])(
      /*#__PURE__*/
      _regenerator["default"].mark(function _callee(json) {
        var publicKey, signingKey;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return _publicKey["default"].decode(json.public_key);

              case 2:
                publicKey = _context.sent;
                signingKey = new _signingKey["default"](null);

                if (!json.hasOwnProperty('signing_key')) {
                  _context.next = 8;
                  break;
                }

                _context.next = 7;
                return _signingKey["default"].decode(json.signing_key);

              case 7:
                signingKey = _context.sent;

              case 8:
                return _context.abrupt("return", Promise.resolve(new ClientDetails(json.client_id, json.api_key_id, json.api_secret, publicKey, json.name, signingKey)));

              case 9:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function decode(_x) {
        return _decode.apply(this, arguments);
      }

      return decode;
    }()
  }]);
  return ClientDetails;
}();

exports["default"] = ClientDetails;
