// e3db-client-interface/dist/types/recordInfo.js
/*!
 * Tozny e3db
 *
 * LICENSE
 *
 * Tozny dual licenses this product. For commercial use, please contact
 * info@tozny.com. For non-commercial use, this license permits use of the
 * software only by government agencies, schools, universities, non-profit
 * organizations or individuals on projects that do not receive external
 * funding other than government research grants and contracts. Any other use
 * requires a commercial license. For the full license, please see LICENSE.md,
 * in this source repository.
 *
 * @copyright Copyright (c) 2018-19 Tozny, LLC (https://tozny.com)
 */
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _signable = _interopRequireDefault(require("./signable"));

/* eslint-disable camelcase */

/**
 * Represents a signed, encrypted documents
 *
 * @property {object} clientMeta
 * @property {object} recordData
 */
var RecordInfo =
/*#__PURE__*/
function (_Signable) {
  (0, _inherits2["default"])(RecordInfo, _Signable);

  function RecordInfo(clientMeta, recordData) {
    var _this;

    (0, _classCallCheck2["default"])(this, RecordInfo);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(RecordInfo).call(this));
    _this.clientMeta = {
      plain: clientMeta.plain,
      type: clientMeta.type,
      user_id: clientMeta.userId,
      writer_id: clientMeta.writerId
    };
    _this.recordData = recordData;
    return _this;
  }
  /* eslint-enabled */

  /**
   * Serialize the object to JSON
   *
   * @returns {string}
   */


  (0, _createClass2["default"])(RecordInfo, [{
    key: "stringify",
    value: function stringify() {
      return JSON.stringify(this.clientMeta) + JSON.stringify(this.recordData);
    }
  }]);
  return RecordInfo;
}(_signable["default"]);

exports["default"] = RecordInfo;
